File: app/src/main/java/com/test/mercaridemo/shoppinglist/repository/ShoppingRepository.java
package com.test.shoppingapp.shoppinglist.repository;


import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.util.Log;

import com.test.networkmodule.api.ShoppingAPI;
import com.test.networkmodule.response.ShopListingResponse;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class ShoppingRepository {

    private static final String TAG = "ShoppingRepository";

    private final ShoppingAPI shoppingAPI;

    private final MutableLiveData<List<ShopListingResponse>> mensShoppingList;
    private final MutableLiveData<List<ShopListingResponse>> womensShoppingList;
    private final MutableLiveData<List<ShopListingResponse>> allShoppingList;

    private final CompositeDisposable compositeDisposable;

    @Inject
    public ShoppingRepository(ShoppingAPI shoppingAPI) {
        this.shoppingAPI = shoppingAPI;
        this.compositeDisposable = new CompositeDisposable();
        this.mensShoppingList = new MutableLiveData<>();
        this.womensShoppingList = new MutableLiveData<>();
        this.allShoppingList = new MutableLiveData<>();
    }

    public void getShoppingListForMen() {
        Disposable disposable = shoppingAPI.getShoppingListForMen()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onMensListReceived, this::onApiError);
        compositeDisposable.addAll(disposable);
    }

    public void getShoppingListForWomen() {
        Disposable disposable = shoppingAPI.getShoppingListForWomen()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onWoensListReceived, this::onApiError);
        compositeDisposable.addAll(disposable);
    }

    public void getShoppingListForAll() {
        Disposable disposable = shoppingAPI.getShoppingListForAll()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onAllListReceived, this::onApiError);
        compositeDisposable.addAll(disposable);
    }

    private void onMensListReceived(List<ShopListingResponse> shopListingResponse) {
        mensShoppingList.setValue(shopListingResponse);
    }

    private void onWoensListReceived(List<ShopListingResponse> shopListingResponse) {
        womensShoppingList.setValue(shopListingResponse);
    }

    private void onAllListReceived(List<ShopListingResponse> shopListingResponse) {
        allShoppingList.setValue(shopListingResponse);
    }

    private void onApiError(Throwable throwable) {
        Log.e(TAG, "Error " + throwable.getMessage(), throwable);
    }

    public LiveData<List<ShopListingResponse>> getMensShoppingList() {
        return mensShoppingList;
    }

    public LiveData<List<ShopListingResponse>> getWomensShoppingList() {
        return womensShoppingList;
    }

    public LiveData<List<ShopListingResponse>> getAllShoppingList() {
        return allShoppingList;
    }

    public void onCleared() {
        compositeDisposable.clear();
    }

}

File: app/src/main/java/com/test/shoppingapp/shoppinglist/repository/ShoppingRepository.kt
package com.test.shoppingapp.shoppinglist.repository


import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.test.networkmodule.api.ShoppingAPI
import com.test.networkmodule.response.ShopListingResponse
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Inject

/**
 * ShoppingRepository takes care of fetching listing response for different categories displayed on home page.
 * Even though ViewModel could do that it's not a good idea to bloat VM with data fetching logic because
 * it could be doing that for many different APIs
 */
class ShoppingRepository @Inject
constructor(private val shoppingAPI: ShoppingAPI) {

    private val mensShoppingList: MutableLiveData<List<ShopListingResponse>> = MutableLiveData()
    private val womensShoppingList: MutableLiveData<List<ShopListingResponse>> = MutableLiveData()
    private val allShoppingList: MutableLiveData<List<ShopListingResponse>> = MutableLiveData()
    val errorData: MutableLiveData<String> = MutableLiveData()

    private val compositeDisposable: CompositeDisposable = CompositeDisposable()

    fun getShoppingListForMen() {
        val disposable = shoppingAPI.shoppingListForMen
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ this.onMensListReceived(it) },
                { this.onApiError(it) })
        compositeDisposable.addAll(disposable)
    }

    fun getShoppingListForWomen() {
        val disposable = shoppingAPI.shoppingListForWomen
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { this.onWomensListReceived(it) },
                { this.onApiError(it) })
        compositeDisposable.addAll(disposable)
    }

    fun getShoppingListForAll() {
        val disposable = shoppingAPI.shoppingListForAll
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { this.onAllListReceived(it) },
                { this.onApiError(it) })
        compositeDisposable.addAll(disposable)
    }

    private fun onMensListReceived(shopListingResponse: List<ShopListingResponse>) {
        mensShoppingList.value = shopListingResponse
    }

    private fun onWomensListReceived(shopListingResponse: List<ShopListingResponse>) {
        womensShoppingList.value = shopListingResponse
    }

    private fun onAllListReceived(shopListingResponse: List<ShopListingResponse>) {
        allShoppingList.value = shopListingResponse
    }

    private fun onApiError(throwable: Throwable) {
        Timber.tag(TAG).e(throwable)
        errorData.value = "Something went wrong!"
    }

    fun getMensShoppingList(): LiveData<List<ShopListingResponse>> {
        return mensShoppingList
    }

    fun getWomensShoppingList(): LiveData<List<ShopListingResponse>> {
        return womensShoppingList
    }

    fun getAllShoppingList(): LiveData<List<ShopListingResponse>> {
        return allShoppingList
    }

    fun onCleared() {
        compositeDisposable.clear()
    }

    companion object {
        private val TAG = "ShoppingRepository"
    }

}
